ledger/node.go

```diff
@@ -0,0 +1,307 @@
+	if key == nil {
+		return buildPage(nil, height, db, nil)
+	}
+	var h hash
+	copy(h[:], key)
+	rawPage, exists := db.updatedNodes.Get(h)
+	rawPage = copy2d(rawPage)
+	if exists {
+		return buildPage(rawPage, height, db, key)
+	}
+	return nil
+}
+
+// this code looks redundant, but this actually is the fastest way to calculate floor(log2(N)) where N <=31
+func heightInPage(i byte) byte {
+	if i >= 15 {
+		return 4
+	} else if i >= 7 {
+		return 3
+	} else if i >= 3 {
+		return 2
+	} else if i >= 1 {
+		return 1
+	} else {
+		return 0
+	}
+}
+
+func (n *node) height() byte {
+	// this is mathematically correct but computationally expensive
+	return n.page.height - heightInPage(n.index)
+}
+
+func (n *node) isShortcut() bool {
+	if n.height()%4 != 0 {
+		return len(n.val) != 0 && n.val[hashLength] == 1
+	}
+	return n.getNextPage().nodes[0].val[0] == 1
+}
+
+// returns true if node is a leaf of the page
+func (n *node) isLeaf() bool {
+	return n.index+1 >= 1<<batchHeight //this is easier than calculating 2^4
+}
+
+func (n *node) getNextPage() *page {
+	if n.nextPage == nil {
+		n.nextPage = retrieveOrBuildPage(n.page.db, n.val, n.height())
+	}
+	return n.nextPage
+}
+
+func (n *node) left() *node {
+	if n == nil {
+		return nil
+	}
+	if n.isLeaf() {
+		np := n.getNextPage()
+		if np == nil {
+			return nil
+		}
+		return np.nodes[1]
+	}
+	return n.page.nodes[leftIndex(n.index)]
+}
+
+func (n *node) right() *node {
+	if n == nil {
+		return nil
+	}
+	if n.isLeaf() {
+		np := n.getNextPage()
+		if np == nil {
+			return nil
+		}
+		return np.nodes[2]
+	}
+	return n.page.nodes[rightIndex(n.index)]
+}
+
+func leftIndex(i byte) byte {
+	result := 2*i + 1
+	if result >= byte(batchLen) {
+		return 1
+	}
+	return result
+}
+
+func rightIndex(i byte) byte {
+	result := 2*i + 2
+	if result >= byte(batchLen) {
+		return 2
+	}
+	return result
+}
+
+func (n *node) makeShortcut(key []byte, val []byte) {
+	// n.left and n.right must be nil
+	// mark n as shortcut node
+	var p *page
+	if n.isLeaf() {
+		p = buildPage([][]byte{}, n.page.height-4, n.page.db, []byte{})
+		n.nextPage = p
+		p.nodes[0].val = []byte{1}
+	} else {
+		n.val = make([]byte, hashLength)
+		n.val = append(n.val, 1)
+		p = n.page
+	}
+	l := node{
+		val:   key,
+		index: leftIndex(n.index),
+		page:  p,
+	}
+	r := node{
+		val:   val,
+		index: rightIndex(n.index),
+		page:  p,
+	}
+	p.nodes[l.index] = &l
+	p.nodes[r.index] = &r
+}
+
+func (n *node) removeShortcut() {
+	if n.isLeaf() {
+		p := n.getNextPage()
+		p.nodes[0].val[0] = 0
+		p.nodes[1] = nil
+		p.nodes[2] = nil
+	} else {
+		n.val[hashLength] = 0
+		n.page.nodes[leftIndex(n.index)] = nil
+		n.page.nodes[rightIndex(n.index)] = nil
+	}
+}
+
+func (n *node) calculateHash(hasher func(data ...[]byte) []byte, defaultHashes [][]byte) []byte {
+	var h []byte
+	if (n.left() == nil || len(n.left().val) == 0) && (n.right() == nil || len(n.right().val) == 0) {
+		//s.deleteOldNode(oldRoot) //TODO
```

Comment:
what is the impact of this todo?

Reply:
Ah, my IDE collapsed this line, and I I couldn't see the comment.   This was a reminder to make sure old nodes get expired.  It is handled by Erase().